### PR TITLE
2073354: Print correct status, when access mode has changed

### PR DESCRIPTION
**Pull request by jirihnidek**
* BZ: https://bugzilla.redhat.com/show_bug.cgi?id=2073354
* Card ID: ENT-4885
* When content access mode has changed from SCA mode to
  entitlement mode, then sub-man should resurrect from this
  situation gracefully, when status command is triggered
* Added new service refresh, because it is necessary to
  refresh certificate not only from refresh command, but
  status needs this functionality in edge cases too
* Refactored code of status command
* Fixed some unit tests
* Added unit tests for status command and for this case

### CODE REVIEW

File: src/rhsmlib/services/refresh.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+log = logging.getLogger(__name__)
+
+
+class Refresh(object):
+    """
+    Class used for refreshing entitlement certificates
+    """
+
+    def __init__(self, cp=None, ent_cert_lib=None):
```
Comment (ptoscano) [inline src/rhsmlib/services/refresh.py:32]: both the new callers of this pass both `cp` and `ent_cert_lib`; to make things easier for now, should we avoid the default arguments & code for the `None` cases?

File: src/subscription_manager/cli_command/status.py [hunk 2]
```diff
@@ -46,38 +47,59 @@ def __init__(self):
...
         else:
+            # When there are no entitlement SCA certificates, but status_id is "disabled", then
+            # it means that content access mode has changed on the server and entitlement certificates
+            # have to be refreshed
+            if service_status["status_id"] == "disabled":
+                refresh_service.refresh()
             if is_simple_content_access(uep=self.cp, identity=self.identity):
+                sca_mode_detected = True
+
+        if sca_mode_detected is True:
```
Comment (ptoscano) [inline src/subscription_manager/cli_command/status.py:95]: comparing a boolean with True is a bit redundant...

File: src/rhsmlib/services/refresh.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+        # SCA or regular mode correctly
+        content_access_mode = inj.require(inj.CONTENT_ACCESS_MODE_CACHE)
+        if content_access_mode.exists():
+            content_access_mode.delete_cache()
+
+        # Remove the release status cache, in case it was changed
+        # on the server; it will be fetched when needed again
+        inj.require(inj.RELEASE_STATUS_CACHE).delete_cache()
+
+        if force is True:
```
Comment (ptoscano) [inline src/rhsmlib/services/refresh.py:65]: comparing a boolean with True is a bit redundant...

File: src/subscription_manager/cli_command/status.py [hunk 4]
```diff
@@ -46,38 +47,59 @@ def __init__(self):
...
+            # When SCA mode was detected using cache or installed SCA entitlement certificates, but status_id
+            # is not "disabled", then it means that content access mode has changed on the server and entitlement
+            # certificates have to be refreshed
+            if service_status["status_id"] != "disabled":
+                log.debug("Found SCA cert, but status ID is not 'disabled'. Refreshing entitlement certs...")
```
Comment (ptoscano) [inline src/subscription_manager/cli_command/status.py:100]: maybe extend this debug like eg like `... not 'disabled' ({service_status["status_id"]}). Refreshing ...`; this way we can know what's the actual status when reading this log entry

File: test/cli_command/test_status.py [hunk 5]
```diff
@@ -12,6 +30,84 @@ class TestStatusCommand(SubManFixture):
...
+        cert_list = [mock_cert]
+        self.cc.entitlement_dir = Mock()
+        self.cc.entitlement_dir.list_with_content_access = Mock(return_value=cert_list)
+        self.cc.entcertlib = Mock()
+        with Capture() as cap:
+            self.cc._do_command()
+        self.assertTrue("Overall Status: Disabled" in cap.out)
+        self.assertTrue("Content Access Mode is set to Simple Content Access." in cap.out)
+        self.assertTrue("This host has access to content, regardless of subscription status." in cap.out)
+        self.assertTrue("System Purpose Status: Disabled" in cap.out)
```
Comment (ptoscano) [inline test/cli_command/test_status.py:61]: `assertIn()` please :)

File: test/cli_command/test_status.py [hunk 6]
```diff
@@ -12,6 +30,84 @@ class TestStatusCommand(SubManFixture):
...
+        mock_cert = Mock()
+        mock_cert.entitlement_type = "entitlement"
+        cert_list = [mock_cert]
+        self.cc.entitlement_dir = Mock()
+        self.cc.entitlement_dir.list_with_content_access = Mock(return_value=cert_list)
+        self.cc.entcertlib = Mock()
+        with Capture() as cap:
+            self.cc._do_command()
+        self.assertTrue("Overall Status: Current" in cap.out)
+        self.assertTrue("System Purpose Status: Matched" in cap.out)
```
Comment (ptoscano) [inline test/cli_command/test_status.py:85]: `assertIn()` please :)

File: test/cli_command/test_status.py [hunk 7]
```diff
@@ -12,6 +30,84 @@ class TestStatusCommand(SubManFixture):
...
+        mock_cert.entitlement_type = CONTENT_ACCESS_CERT_TYPE
+        cert_list = [mock_cert]
+        self.cc.entitlement_dir = Mock()
+        self.cc.entitlement_dir.list_with_content_access = Mock(return_value=cert_list)
+        self.cc.entcertlib = Mock()
+        # sub-man should be able to resurrect from this situation
+        with Capture() as cap:
+            self.cc._do_command()
+        self.assertTrue("Overall Status: Current" in cap.out)
+        self.assertTrue("System Purpose Status: Matched" in cap.out)
```
Comment (ptoscano) [inline test/cli_command/test_status.py:110]: `assertIn()` please :)

File: src/subscription_manager/cli_command/refresh.py [hunk 8]
```diff
@@ -17,6 +17,7 @@
 import logging
 import os
 
+from rhsmlib.services.refresh import Refresh
 import rhsm.connection as connection
 import subscription_manager.injection as inj
```
Comment (ptoscano) [inline src/subscription_manager/cli_command/refresh.py:22]:
```
./src/subscription_manager/cli_command/refresh.py:22:1: F401 'subscription_manager.injection as inj' imported but unused
```

**Review by ptoscano — CHANGES_REQUESTED** [review]
It seems mostly OK to me, there are few minor things to change/fix.

File: src/rhsmlib/services/refresh.py (hunk 1 above)
Comment (jirihnidek) [reply src/rhsmlib/services/refresh.py:32]: Yes, new callers pass something, but future callers needn't. Default arguments makes usage of this class easier.

File: src/subscription_manager/cli_command/refresh.py (hunk 8 above)
Comment (jirihnidek) [reply src/subscription_manager/cli_command/refresh.py]: Done.

File: test/cli_command/test_status.py (hunk 7 above)
Comment (jirihnidek) [reply test/cli_command/test_status.py:110]: Done

File: test/cli_command/test_status.py (hunk 6 above)
Comment (jirihnidek) [reply test/cli_command/test_status.py:85]: Done

File: test/cli_command/test_status.py (hunk 5 above)
Comment (jirihnidek) [reply test/cli_command/test_status.py:61]: Done

File: src/subscription_manager/cli_command/status.py (hunk 2 above)
Comment (jirihnidek) [reply src/subscription_manager/cli_command/status.py:95]: Well, I don't like the this approach: `if something:` It could lead to bad results and I just will never use it despite it could look like redundant.

File: src/subscription_manager/cli_command/status.py (hunk 4 above)
Comment (jirihnidek) [reply src/subscription_manager/cli_command/status.py:100]: Done

**Review by ptoscano — APPROVED** [review]
Thanks!
